src/client/websocket/WebSocketShard.js

```diff
@@ -761,6 +856,63 @@ class WebSocketShard extends EventEmitter {
     }
   }
 
+  /**
+   * Forcefully closes the WebSocket when {@link WebSocketShard#destroy} could not close it properly.
+   * @private
+   * @returns {Promise<void>}
+   */
+  async handleZombieConnection() {
+    // Continuing from step 2 from the destroy method.
+
+    // Step 3 Check if the connection was closed and readyState is at CLOSED. If not close it.
+    if (this.connection?.readyState === WebSocket.CLOSING || this.connection?.readyState === WebSocket.OPEN) {
+      this.debug(
+        `[WS Destroy] Step 3: WebSocket Still Connected: trying to terminate. | WS State: ${
+          CONNECTION_STATE[this.connection.readyState]
+        }`,
+      );
+
+      // Trying to close WebSocket with terminate..
+      try {
+        this.connection.terminate();
+      } catch {
+        // No op.
+      }
+
+      // Wait for 400ms for ws to close.
+      await sleep(400);
+      this.debug(`[WebSocket] Final Close Check. | WS State: ${CONNECTION_STATE[this.connection.readyState]}`);
+      if (this.connection.readyState === WebSocket.CLOSING || this.connection.readyState === WebSocket.OPEN) {
+        this.debug(
+          // eslint-disable-next-line max-len
+          `[WebSocket] Connection still stuck at Closing: calling a manual destroy of the socket to close and reconnect. | WS State: ${
+            CONNECTION_STATE[this.connection.readyState]
+          }`,
+        );
+
+        /*
+         * This is an important step to deal with zombie connections wherein shards never reconnect
+         * after a 4009 closeCode due to the WebSocket being stuck at the CLOSING ready state.
+         * Check the issue https://github.com/discordjs/discord.js/issues/7450
+         *
+         * The _socket.destroy() method ensures that no more I/O activity happens on this socket.
+         * Destroys the stream and closes the connection. Refer: https://nodejs.org/api/net.html#socketdestroy
+         * _socket.destroy() is also being invoked in ws.terminate() method.
+         * _socket.destroy() emits the close event and makes the readyState to CLOSED.
+         */
+
+        // manual destory
+        this.connection._socket.destroy();
```

Comment:
I'm unsure what to feel about this, since it's an object managed by a third-party package, and also a private property.

Also, typo on `destory`.

Reply:
Since there is no proper answer from ws package on "why" readyState is stuck at `CLOSING` forever. I just had to directly destroy the socket for them and emit the event manually. Even tho the code for `ws.terminate()` is pretty much similar.

All these changes and codes are in production right now and I have seen 0 downtimes. except a few 4009 where some session cannot be resumed

Reply:
> All these changes and codes are in production right now

It's more referring to the fact that because it's private, the creators of the `ws` package may change it without prior notice, e.g. making the field a true-private one.

In case that happens, I believe we should be prepared for it.

wdyt @vladfrangu?

Reply:
Imo, the proper fix for a 4009 is to not care anymore about the connection. Remove all listeners, terminate the connection and start a-new, no waiting or anything. If we get in a state where we emit a 4009, no point to wait for something that won't happen 👍

Reply:
> Imo, the proper fix for a 4009 is to not care anymore about the connection. Remove all listeners, terminate the connection and start a-new, no waiting or anything. If we get in a state where we emit a 4009, no point to wait for something that won't happen 👍


But as far I have seen with my bot, There are many times where the bot resumes the session after a 4009 with this pr changes ie 0 downtimes, but if we do what you said, > Remove all listeners, terminate the connection, and start a-new.
I think this will cause the bot to have tiny downtime.

Reply:
you'll always have downtime, what matters is trying to resume afterwards

Reply:
> you'll always have downtime, what matters is trying to resume afterwards

So we should directly clear everything and make a new connection for 4009 without even trying to resume the session even once?

Reply:
I also want to ask if anyone knows why only one of my shards has this issue? and it[4009] only appears on it. The rest of my 15 shards are fine.

Reply:
> > you'll always have downtime, what matters is trying to resume afterwards
> 
> So we should directly clear everything and make a new connection for 4009 without even trying to resume the session even once?

No. I said you should clean the ws listeners (this.clearConnectionListeners or w/e its called), call this.connection.terminate() and simply not care if it works or doesn't and re-schedule the shard to connect. Thus, you get to attempt a resume, if it works 🎉, if it doesn't oh well, identify